bfasst/transform/spydrnet_injector.py

```diff
@@ -0,0 +1,199 @@
+from bfasst.status import Status, TransformStatus
+
+
+class SpydrNetErrorInjector(TransformTool):
+    """Tool to inject errors into a netlist"""
+
+    success_status = Status(TransformStatus.SUCCESS)
+    TOOL_WORK_DIR = "spydrnet_injector"
+
+    def __init__(self, cwd, design) -> None:
+        super().__init__(cwd, design)
+        self.clean_netlist = sdn.parse(self.design.reversed_netlist_path)
+        self.hierarchical_luts = []
+        self.flattened_luts = None
+
+    def inject(self, error_type):
+        """Injects an error into the netlist of the given type"""
+        if error_type == "BIT_FLIP":
+            self.inject_bit_flip()
+        elif error_type == "WIRE_SWAP":
+            self.inject_wire_swap()
+        else:
+            return Status(TransformStatus.ERROR, "Invalid error type")
+        return Status(TransformStatus.SUCCESS)
+
+    def inject_bit_flip(self):
+        """Injects a bit flip error into the netlist"""
+        num_luts = self.pick_luts_from_netlist()
+        self.flatten_luts()
+        lut_number = randrange(num_luts)
+        lut_size = self.get_lut_size(lut_number)
+        bit_number = randrange(lut_size)
+        self.flip_bit(lut_number, bit_number)
+        self.compose_corrupt_netlist()
+
+    def pick_luts_from_netlist(self):
+        """Calculates the number of LUTs in the netlist"""
+        num_luts = 0
+        for library in self.clean_netlist.libraries:
+            for definition in library.definitions:
+                if "LUT" in definition.name.upper():
+                    num_luts += len(definition.references)
+                    self.hierarchical_luts.append(definition.references)
+        return num_luts
+
+    def flatten_luts(self):
+        """Flattens the LUTs into a single list"""
+        self.flattened_luts = [lut for sublist in self.hierarchical_luts for lut in sublist]
+
+    def get_lut_size(self, lut_number):
+        """Gets the size of the LUT init string for the given LUT"""
+        lut = self.flattened_luts[lut_number]
+        init_string = lut.data["VERILOG.Parameters"]["INIT"].upper()
+        return int(init_string.split("'")[0])
+
+    def flip_bit(self, lut_number, bit_number):
+        """Flips the bit at the given index"""
+        lut = self.flattened_luts[lut_number]
+        lut_properties = lut.data["VERILOG.Parameters"]
+
+        config_string_prefixed = lut_properties["INIT"].upper()
+        config_string_digits = config_string_prefixed.split("H")[1]
+        lut_size = self.get_lut_size(lut_number)
+        config_as_binary = bin(int(config_string_digits, 16))[2:].zfill(lut_size)
+
+        if config_as_binary[bit_number] == "1":
+            config_as_binary = (
+                config_as_binary[:bit_number] + "0" + config_as_binary[bit_number + 1 :]
+            )
+        else:
+            config_as_binary = (
+                config_as_binary[:bit_number] + "1" + config_as_binary[bit_number + 1 :]
+            )
```

Comment:
Rather than flipping a bit using the string, the usual way to flip a random bit would be:

```
INIT_flipped = INIT ^ (1 << randint)
```

Reply:
...also, to make this more robust (for example the INIT isn't guranteed to be provided in 'h hex format), you could use this function: https://github.com/byuccl/bfasst/blob/main/bfasst/utils.py#L90

It's a nested function so you would need to pull it out to use it.  We should probably name it `verilog_literal_int_val` or something like that.

Reply:
Done